Page Templates: Fix Blank title insertion, some code cleanup

marekhrabe opened the pull request:
Fixing the Blank title bug was connected to some other parts of code so I did a bit of a cleanup along the way.

#### Changes proposed in this Pull Request

* fix the bug when selecting a blank template would insert the title "Blank"
* removed long unused code for `onFocus` (which had nothing to do with a focus)
* removed multiple sources of truth and extraneous passing for template titles between components
* select the first template by default (Blank)

#### Testing instructions

* use inserter as normal, try switching between templates
* insert template with content - everything should work
* insert "Blank" - our plugin should close and you should see nothing being inserted into the editor - no title, no content

Fixes #36060

marekhrabe committed:
fix blank selection, little refactor

marekhrabe committed:
use 'previewedTemplate' as a name

marekhrabe committed:
fix block passing to selector

obenland reviewed (approved):
Tested locally. Looks good to me! :shipit:

getdave reviewed (changes requested):
Nice work! Glad this is fixed. I tested this and confirmed that:

✅ The title (ie: "blank") is not inserted when selecting the Blank template.
✅ The correct Template title _is_ inserted when selecting a valid Template other than the blank one.

My only other concerns are:

1. Have / will you updated the tests (shout if you need help with setup...etc)?
2. Are we merging too many changes into this PR. Lots of small things add up and are also making this harder to review code-wise.

Thanks for working on this 👍

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/template-selector-item.js:
```diff
@@ -54,8 +53,7 @@ const TemplateSelectorItem = props => {
 			value={ value }
-			onMouseEnter={ () => onFocus( value, label ) }
-			onClick={ () => onSelect( value, label ) }
+			onClick={ () => onSelect( value ) }
```
from PR description:

> removed long unused code for onFocus (which had nothing to do with a focus)

marekhrabe replied:
I was standing between the choice to either update the function call to keep it working (the same way as `onClick`) or remove it because we don't use it at all. I chose the 🔥method but happy to revert if you want the fn to stay for some reason.

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -33,16 +33,22 @@ const {
+		// Select the first template automatically.
+		if ( hasTemplates ) {
+			this.state.previewedTemplate = get( props.templates, [ 0, 'slug' ] );
+			this.state.templatesBySlug = keyBy( props.templates, 'slug' );
```
Do you think that it's completely needed to populate the state with the `templatesBySlug` object? It seems to be used only to get the template title. What about...

```es6
	getTitleBySlug( slug = this.state.previewedTemplate ) {
		return get( find( this.props.templates, { slug } ), 'title', '' );
	}
```

marekhrabe replied:
While the `find` solution definitely works, it requires to always loop over the array for each call of that function which we do twice per render. Having the dictionary in state created by `keyBy` makes this loop happen once per page view and any further lookup is instantaneous while keeping a really low memory impact (remember objects are not copies but references here). 

The only disadvantage is that `templatesBySlug` won't update if we get re-rendered with a different set of `templates` but this is true for our whole plugin so I didn't bother at this point (also because that situation is not possible to be triggered with our current code).

retrofox replied:
Ok, got it. Already getting into very detailed issues, make sense just store the title instead of the whole template object? My concern provably unfounded is filling the state with duplicated data.
Take this comment much as a teaching session rather than a feedback request. ;-)

marekhrabe replied:
We are definitely in a theoretical area here but I like it! Normally, I would have implemented all this props/state mess as a Redux store and in those, it is a well established pattern to use two data structs to store data like this:

- `templatesBySlug` - an object keyed by the slug with values of the full template objects. an example:

```
templatesBySlug = {
  blank: { title: 'Blank', … },
  about: { title: 'About', … },
  …
}
```

- and the other part `templateList` - an array of slugs that are supposed to be displayed (also explicitly stating their order). an example:

```
templateList = [ 'blank', 'about', 'contact', … ]
```

I think we are slowly getting there and I expect us to switch to Redux at some point (with @wordpress/data) so just keeping it here for an inspiration.

Regarding storing just titles, the difference is that those would be stored as copies in memory (doubling the usage) as that's what JS does for scalar types. On the other hand, when storing objects, they are not copies but only references.

In case of titles, the memory impact is extremely tiny even for copies (they are short strings and there is a small number of them). I'm happy to switch to that approach for now as we don't access anything other than titles.

retrofox replied:
I really appreciate it. :thanks:

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -134,16 +138,12 @@ class PageTemplateModal extends Component {
-			return [];
-		}
+	getBlocksBySlug( slug = this.state.previewedTemplate ) {
+		return get( this.state.blocksBySlug, [ slug ], [] );
```
:100:

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -134,16 +138,12 @@ class PageTemplateModal extends Component {
-		if ( ! this.state.blocks.hasOwnProperty( slug ) ) {
-			return [];
-		}
+	getBlocksBySlug( slug = this.state.previewedTemplate ) {
```
Not 100% sure about renaming the method here, since `templateSlug` describes that it refers to the template slug. It worths in this context, imho.

marekhrabe replied:
ok, it's using the longer wording from now on 👍

retrofox reviewed (commented):
It works as expected. I got similar concerns as @getdave 

> Are we merging too many changes into this PR. Lots of small things add up and are also making this harder to review code-wise.

We're going to introduce many changes that maybe aren't completely needed to fix the original issue, and it could impact on other patches. Anyway, it's part of our job ;-).

Left some comments, which I think make worth take a look, especially this one: 
https://github.com/Automattic/wp-calypso/pull/36270#pullrequestreview-293098588
I'd like to know your opinion before to approve this one. 🙇

marekhrabe commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -63,35 +69,30 @@ class PageTemplateModal extends Component {
 		}
 
+		// Show loading state.
+		this.setState( {
```
No, it has been introduced as part of the ability to side load images. It has only been moved in this PR (and prepended with a comment).

marekhrabe commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -193,10 +193,10 @@ class PageTemplateModal extends Component {
 						onClick={ this.handleConfirmation }
 					>
-						{ sprintf( __( 'Use %s template', 'full-site-editing' ), this.state.title ) }
+						{ sprintf( __( 'Use %s template', 'full-site-editing' ), this.getTitleBySlug() ) }
```
I was actually in the same boat - I don't like that much but I wanted to follow what was in the code. I would much rather remove the default fallback and always pass slug.

marekhrabe commented:
Just to clarify about the PR scope, all those things are deeply connected. I initially started just by fixing the "Blank" title problem but it was just another bandaid on top of previous ones. So I went through the rabbit hole - the template title was duplicated in state and was passed in both directions (from parent to child but sadly also from child to parent). So I removed that explicit passing in favor of always reading it from the template object. Making a getter fn for the title allowed me to use in both the UI and the template insertion. That however lead to the issue when initial state of the modal said "Use undefined template" because no template is initially selected. So I've added the automattic selection of the first template (which happens to be Blank). 

I'm always in favor of smaller PRs but I felt that in this case if we don't do those together, their separate PRs would all be half-functioning without each other and even less testable. I believe the diff size is not too terrible in this case but I'll aim for smaller ones next time and spend more time trying to figure out how to split more.

retrofox commented:
> I believe the diff size is not too terrible in this case but I'll aim for smaller ones next time and spend more time trying to figure out how to split more.

No worries at all, completely sure we understand this situation. A question that I'm interesting to know here is why we are storing the templates (by slug) in the state when we already have them into the props. since it doesn't seem to be necessary. https://github.com/Automattic/wp-calypso/pull/36270#pullrequestreview-293098588, but don't take it as a blocker and go ahead!

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -115,13 +117,15 @@ class PageTemplateModal extends Component {
-	previewTemplate = ( slug, title ) => {
-		this.setState( { slug, title } );
+	previewTemplate = slug => {
+		// Immediately confirm and close when 'blank' is selected.
```
Keep in mind this PR was recently closed.

marekhrabe committed:
be more consistent using templateSlug, explicit slug passing to getters

marekhrabe commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -171,17 +171,17 @@ class PageTemplateModal extends Component {
+								blocks={ this.getBlocksBySlug() }
 								viewportWidth={ 960 }
-								title={ this.state.title }
+								title={ this.getTitleBySlug() }
```
now always explicit

marekhrabe committed:
store only titles for fast lookup

marekhrabe committed:
remove test for deleted functionality

marekhrabe commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -33,16 +33,22 @@ const {
-		this.state.isOpen = ! isEmpty( props.templates );
+		const hasTemplates = ! isEmpty( props.templates );
+		this.state.isOpen = hasTemplates;
+		// Select the first template automatically.
```
> wrapped up in a PR which is concerned with blank template titles.

I left longer explanation as a comment in the main PR thread. tldr: when there was no template selected (that was the initial state) and titles were no longer duplicated in the state, it lead to the confirmation button saying "Use undefined template" as it referenced non-existing template. This change, although not seeming related, made sure we don't need to add more complexity to the button rendering logic and there will always be a template selected.

marekhrabe commented:
I hope I've addressed most of the feedback now. I left a few conversations opened where I'm not sure you are ok with proceeding. Let me know!

retrofox commented:
- [x] fix the bug when selecting a blank template would insert the title "Blank"
- [x] removed long unused code for onFocus (which had nothing to do with a focus)
- [x] removed multiple sources of truth and extraneous passing for template titles between components
- [x] select the first template by default (Blank)

<img width="556" alt="Screen Shot 2019-09-26 at 10 13 48 AM" src="https://user-images.githubusercontent.com/77539/65690937-63e81f00-e046-11e9-9d49-74c6efe67b69.png">

retrofox reviewed (approved):
Thanks, Marek for this PR. Improves the code considerably 👍

marekhrabe committed:
allow 'blank' to be previewed
